main.go

```diff
@@ -34,23 +63,70 @@ func main() {
 		},
 	}
 
-	proxy.Transport = http.DefaultTransport
+	if *oidcConfig.Enabled {
+		provider, err := oidc.NewProvider(context.Background(), *oidcConfig.IssuerUrl)
+		if err != nil {
+			panic(err)
+		}
+
+		var cfg = clientcredentials.Config{
+			ClientID:     *oidcConfig.ClientId,
+			ClientSecret: *oidcConfig.ClientSecret,
+			TokenURL:     provider.Endpoint().TokenURL,
+		}
+
+		if *oidcConfig.Audience != "" {
+			cfg.EndpointParams = map[string][]string{
+				"audience": {*oidcConfig.Audience},
+			}
+		}
+
+		proxy.Transport = &oauth2.Transport{
+			Source: cfg.TokenSource(context.Background()),
+			Base:   http.DefaultTransport,
+		}
+	} else {
+		proxy.Transport = http.DefaultTransport
+	}
 
 	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
-		err := remotewrite.ValidateRequest(r)
+		// extract the remote write request from the http request
+		remoteWriteRequest, err := remotewrite.DecodeWriteRequest(r)
 		if err != nil {
+			log.Printf("error decoding remote write request: %v", err)
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+
+		log.Printf("remote write request received")
+
+		// validate the cluster ids contained in the remote write request
+		err = remotewrite.ValidateRequest(remoteWriteRequest)
+		if err != nil {
+			log.Printf("error validating the remote write request: %v", err)
 			w.WriteHeader(http.StatusForbidden)
-		} else {
-			proxy.ServeHTTP(w, r)
 		}
+
+		// copy the remote write request back onto the http request
+		err = populateRequestBody(remoteWriteRequest, r)
```

Comment:
This is needed because the body has been read already?

Reply:
Yes, exactly

Reply:
It is reasonable. In an ideal scenario, we would have wished to not have to rewrite the req body again but we do not have a way out.